File: apps/ios/Wordocious/Sources/ReviewPrompter.swift
import StoreKit
import SwiftUI
#if canImport(UIKit)
import UIKit
#endif

/// App Store review prompt with deliberate timing: only after a WIN while the
/// player is on a streak of 3+, at most once per app version, delayed a beat so
/// it never competes with the confetti. Never fires on a loss or at launch —
/// asking at a high point is worth real stars.
enum ReviewPrompter {
    @MainActor
    static func maybeAskAfterWin() {
        guard (AuthService.shared.profile?.currentStreak ?? 0) >= 3 else { return }
        let version = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "0"
        let key = "review-asked-v\(version)"
        guard !UserDefaults.standard.bool(forKey: key) else { return }
        UserDefaults.standard.set(true, forKey: key)
        #if canImport(UIKit)
        // Let the win celebration land first.
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.5) {
            guard let scene = UIApplication.shared.connectedScenes
                .first(where: { $0.activationState == .foregroundActive }) as? UIWindowScene else { return }
            SKStoreReviewController.requestReview(in: scene)
        }
        #endif
    }
}

File: apps/android/app/src/main/kotlin/com/wordocious/app/data/ReviewPrompter.kt
package com.wordocious.app.data

import android.app.Activity
import android.content.Context
import com.google.android.play.core.review.ReviewManagerFactory
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * Play in-app review with deliberate timing: only after a WIN while the player
 * is on a streak of 3+, at most once per version, delayed a beat so it never
 * competes with the confetti. Never on a loss or at launch. (Play itself also
 * quota-limits how often the sheet actually shows.)
 */
object ReviewPrompter {
    fun maybeAskAfterWin(context: Context) {
        val streak = AuthService.profile.value?.currentStreak ?: 0
        if (streak < 3) return
        val version = runCatching {
            context.packageManager.getPackageInfo(context.packageName, 0).versionName
        }.getOrNull() ?: "0"
        val prefs = context.getSharedPreferences("review", Context.MODE_PRIVATE)
        val key = "asked-v$version"
        if (prefs.getBoolean(key, false)) return
        prefs.edit().putBoolean(key, true).apply()

        val activity = context as? Activity ?: return
        CoroutineScope(Dispatchers.Main).launch {
            delay(2500) // let the win celebration land first
            runCatching {
                val manager = ReviewManagerFactory.create(activity)
                manager.requestReviewFlow().addOnSuccessListener { info ->
                    manager.launchReviewFlow(activity, info)
                }
            }
        }
    }
}
